Refactor client logic

not-an-aardvark opened the pull request:
Now the client doesn't have to make any further requests once a page is loaded, so everything loads much faster. Also, there are a bunch more unit tests, and a few more serverside checks for things (e.g. whether a user is eligible for a particular flair)

not-an-aardvark commented on api/controllers/AuthController.js:
```diff
@@ -25,19 +25,33 @@ module.exports = {
-        sails.warn("Failed login for /u/" + user.name + ": invalid session state");
-        return res.forbidden();
+        console.log("Failed login for /u/" + user.name + ": invalid session state");
+        return res.view(403, {error: 'You have an invalid session state. (Try logging in again.)'});
```
Just refactored some stuff here. Also, people will be told when they have an invalid session state or are banned, rather than just seeing the default 403 page.

not-an-aardvark commented on api/controllers/AuthController.js:
```diff
@@ -25,49 +25,67 @@ module.exports = {
+        }
+        if (login_info.validation !== req.session.validation) {
+          console.log("Failed login for /u/" + user.name + ": invalid session state");
+          return res.view(403, {error: 'You have an invalid session state. (Try logging in again.)'});
```
Now users will be informed when they have an invalid session state or when they're banned, rather than just seeing the default 403 page.

not-an-aardvark commented on api/controllers/AuthController.js:
```diff
@@ -25,49 +25,67 @@ module.exports = {
+          return finishLogin();
+        }
+      } catch (err) {
+        return res.serverError(err);
```
I put this function in a try/catch since it's async, so that accounts for most of this diff.

not-an-aardvark commented on api/controllers/FlairController.js:
```diff
@@ -5,29 +5,28 @@ var _ = require("lodash");
+      return res.ok(await Application.create(appData));
+    } catch (err) {
+      return res.serverError(err);
+    }
```
Refactored this and made it async. Also, we now verify that a user is qualified for a particular flair when they apply for it.

not-an-aardvark commented on api/controllers/FlairController.js:
```diff
@@ -130,88 +129,93 @@ module.exports = {
+        return res.ok(req.user);
+      });
+    } catch (err) {
+      return res.serverError(err);
```
This is all the same as before, just dumped it into a try/catch since it's async.

not-an-aardvark commented on api/controllers/HomeController.js:
```diff
@@ -31,18 +31,23 @@ module.exports = {
-        res.view('404', {data: {user: req.params.user, error: "User not found"}});
+  reference: async function(req, res) {
+    try {
+      return res.view({refUser: await Users.get(req.user, req.params.user)});
```
We use the `user` ejs local to describe to describe the currently authenticated user, as we did before. However, `user` no longer contains the user's references. Instead, we use `refUser` for the user whose references are being viewed. (On the index page, `refUser.name === user.name`.

YaManicKill replied:
I'm a little bit confused as to the Users.get though, why are we sending it a user and a username? That's weird.

YaManicKill replied:
Wait, just noticed it's Users, not User. Saw your new service. Cool cool.

not-an-aardvark commented on api/controllers/ReferenceController.js:
```diff
@@ -195,18 +194,16 @@ module.exports = {
+      return await References.approve(ref, req.allParams().approve);
+    } catch (err) {
+      return res.serverError(err);
+    }
```
Refactored this, it's now async. I also added an `isApprovable` function in one of the services. It's mainly used to make the ejs files a bit more compact, but we also use it here to verify that the request to approve something is valid.

not-an-aardvark commented on api/controllers/ReferenceController.js:
```diff
@@ -249,13 +246,12 @@ module.exports = {
+      return res.ok(await Flairs.getFlairs());
+    } catch (err) {
+      return res.serverError(err);
+    }
```
I kept this function since it's nice to have a working api, but this is no longer used, since the flairs are sent as an EJS local when viewing a page anyway. I also refactored this to make it async.

not-an-aardvark commented on api/controllers/UserController.js:
```diff
@@ -81,72 +81,15 @@ module.exports = {
-          res.ok(req.user);
-        });
-      });
-  },
```
I removed /user/mine, instead we one can just use `/user/get/<my username>`.

YaManicKill replied:
Yeah, good call. Unnecessary now that we pass the user straight to the ejs.

not-an-aardvark commented on api/controllers/UserController.js:
```diff
@@ -81,72 +81,15 @@ module.exports = {
-
-    Promise.all(promises).then(function () {
-      return res.ok(user);
-    });
```
Moved this to a service.

not-an-aardvark commented on api/controllers/UserController.js:
```diff
@@ -194,107 +137,111 @@ module.exports = {
+        content: "Banned /u/" + req.params.username
+      }).exec(function () {});
+    } catch (err) {
+      return res.serverError(err);
```
I put this in a try-catch since it's async. Everything else is the same as before.

not-an-aardvark commented on api/controllers/UserController.js:
```diff
@@ -310,13 +257,12 @@ module.exports = {
+      return res.ok(await Users.getBannedUsers());
+    } catch (err) {
+      return res.serverError(err);
+    }
```
This is no longer used by the client, but it's nice to have as an api feature. Anyway, it just calls a service now.

not-an-aardvark commented on api/policies/bearerAuth.js:
```diff
@@ -1,7 +0,0 @@
-
-  return passport.authenticate('reddit', { session: false })(req, res, next);
-
-};
```
We don't use this and it's a bit misleading, so I removed it.

not-an-aardvark commented on api/policies/passport.js:
```diff
@@ -4,11 +4,21 @@ module.exports = function (req, res, next) {
+        next();
+      } catch (err) {
+        return res.serverError(err);
+      }
```
This is where all the ejs variables are set.

not-an-aardvark commented on api/policies/sessionAuth.js:
```diff
@@ -8,27 +8,15 @@
+  if (req.user) {
+    if (req.user.banned) {
+      req.logout();
+      return res.view(403, {error: "You have been banned from FlairHQ"});
```
Made this a bit smaller and simpler to read.

not-an-aardvark commented on api/services/Flairs.js:
```diff
@@ -116,26 +116,24 @@ exports.getFlairTextForSVEx = function (user) {
-    applicationFlair = exports.getFlair(applicationFlair, allflairs);
-  }
-  if (!user || !user.references || !applicationFlair || exports.userHasFlair(user, applicationFlair) || exports.applied(user, applicationFlair)) {
+exports.canUserApply = function (refs, applicationFlair, currentFlairs) {
```
I changed this so that it can be called properly from both the client and the server.

not-an-aardvark commented on assets/sharedClientFunctions.js:
```diff
@@ -1,149 +0,0 @@
-      return flairService.formattedRequirements(flair, $scope.flairs);
-    };
-  }
-};
```
I moved all these functions to `userCtrl.js` since `$scope` functions are accessible from other controllers anyway.

YaManicKill replied:
Ah right, no. Please don't do this. This means that we rely on specific controllers existing, meaning if we remove this control (or change it) then we also have to change something in another controller. Bad plan.

not-an-aardvark replied:
Alright. In that case, what should we do with the shared functions? There are a few options that I can think of:
1. Have a helper function in another file to add them to the scope, and call that function from *all* the controllers (this is what we were doing before, which was redundant since this helper function only needs to be called once).
2. Have a helper function in another file, but only call the helper function from one of the controllers.
3. Explicitly define all the functions in a controller that's always active (what I was doing here).
4. Add the functions to `$scope` in app.js using [run()](https://docs.angularjs.org/api/ng/type/angular.Module#run). This isn't a bad option, but it still causes functions to end up defined even if they're not needed. It also creates a strange situation where some of our functions are defined in controllers, and some are defined in app.js.

Personally I think 3 or 4 are best.

YaManicKill replied:
So, if we had everything using the controlleras syntax (which we will move to eventually) then we would do 1, because the $scope isn't the same in the controllers. 

Now, if we are only accessing those things in the html, and not the controller itself (I can't remember, it's been a while) then we would only need to define it in 1 controller, and then do

user.sharedFunction (because we would have UserController as user) 

But if we use the code in a controller, we define it there. Because otherwise we will end up with issues. 

Have a read of [this](https://daveceddia.com/sharing-data-between-controllers-best-practice-use-a-service/).

not-an-aardvark replied:
I didn't end up using a service because a lot of the shared functions require other scope variables. But I moved everything back to the shared file, so now it's not tied to a single controller anymore.

not-an-aardvark commented on assets/views/home/references.ejs:
```diff
@@ -19,39 +19,13 @@
-          <span class="glyphicon glyphicon-trash"></span>
-        </a>
-      </td>
+      <%- partial('row.ejs') %>
```
I got rid of some of the repetition in `references.ejs`, now rows for trades are abstracted out.

not-an-aardvark commented on assets/views/layout.ejs:
```diff
@@ -25,7 +25,13 @@
+    (typeof flairs !== 'undefined' ? 'flairs:' + JSON.stringify(flairs) + ',' : '') +
+    (typeof flairApps !== 'undefined' ? 'flairApps:' + JSON.stringify(flairApps) + ',' : '') +
+    (typeof query !== 'undefined' ? 'query:' + JSON.stringify(query) + ',' : '')
+  %>})">
```
Here's where the EJS variables get injected into angular variables. Essentially, the `$scope.init()` function is called with an object containing the existing EJS variables.

YaManicKill replied:
I do hate this, but it makes other things nicer, and I don't think there's a nicer way to do it. So that's fine.

not-an-aardvark commented on config/blueprints.js:
```diff
@@ -44,7 +44,7 @@ module.exports.blueprints = {
   ***************************************************************************/
 
-  // actions: true,
+  actions: false,
```
I got rid of auto-mapped routes, all our routes are mapped explicitly now.

YaManicKill replied:
\o/

not-an-aardvark commented on test/unit/data/referenceFactory.js:
```diff
@@ -0,0 +1,53 @@
+    }
+    return refs;
+  }
+};
```
Generates random, mostly-valid references for use in unit testing. Certain constraints can be placed on it (e.g. you can tell it to generate only `event` trades, or only SVExchange references).

not-an-aardvark commented on test/unit/data/standardFlairInfo.json:
```diff
@@ -0,0 +1,54 @@
+    "smartribbon":"Smart Ribbon",
+    "toughribbon":"Tough Ribbon"
+  }
+}
```
This reminds me: Do we really need a separate database collection for flairs? They change very infrequently -- it seems like it would be easier just to have them hard-coded into a json file, and have the formatted names explicitly stored with them rather than relying on parsing the regular names.

YaManicKill replied:
It was nice when we changed it for anyone to be able to change what we had.

not-an-aardvark commented on test/unit/services/Flairs.test.js:
```diff
@@ -1,88 +1,155 @@
+    var userFlairs = Flairs.getUserFlairs(users.default_flair_user, stdFlairInfo.flairs);
+    var refs = refFactory.getRefs(10, {type: 'bank'});
+    assert(!Flairs.canUserApply(refs, stdFlairInfo.flairs.pokeball, userFlairs), 'Error: Bank trades are counted for flair');
+  });
```
Added some more tests, you can run `grunt test` to see them.

YaManicKill replied:
Wow, that's a lot of tests. Dunno if I can be bothered reading them.

not-an-aardvark commented on api/services/Flairs.js:
```diff
@@ -253,3 +278,43 @@ exports.getSimilarBannedFCs = function (fc) {
+  return Flair.find({});
+};
+
+exports.refreshAppClaim = function (ref, mod_username) {
```
Whenever a mod clicks a ref link, their client sends a POST request to the server containing info on the ref that they clicked. The server interprets this ref and runs a query to determine which application the mod was working on. If any applications match, they are marked as 'claimed' by the mod, and remain so for 5 minutes.

YaManicKill commented on api/controllers/AuthController.js:
```diff
@@ -25,49 +25,67 @@ module.exports = {
      * might fail to complete the login and then be confused when they get redirected somewhere unexpected the next time they visit the site. */
     var login_info = {type: req.query.loginType, redirect: req.query.redirect || '/', validation: crypto.randomBytes(32).toString('hex')};
     req.session.validation = login_info.validation;
-    var auth_data = {state: JSON.stringify(login_info), duration: 'permanent', failureRedirect: '/login'};
```
Why no failure redirect here?

not-an-aardvark replied:
Was the failure redirect actually doing anything? I was always getting redirected to the 403 page on failure anyway.

YaManicKill commented on api/controllers/AuthController.js:
```diff
@@ -25,49 +25,67 @@ module.exports = {
+    if (req.query.loginType === 'mod') {
+      auth_data.duration = 'permanent'; //Mods have a permanent access token. Scope is not specified here, so the scope from config/express.js is used.
+    } else {
+      auth_data.scope = 'identity'; //Regular users only need a temporary access token with reduced scope.
```
So, do we not call anything for users at all after initial login?

YaManicKill replied:
Also, where is the scope for the mod?

not-an-aardvark replied:
Right. now that we only request `identity` scope from users there's not much to call anyway.

The scope for the mod is that list in `config/express.js`. This is like what we had before, where the the normal-user login overrides the `default` scopes we have listed there.

YaManicKill commented on api/controllers/FlairController.js:
```diff
@@ -5,122 +5,65 @@ var _ = require("lodash");
-      });
-    });
+      var applicationFlair = allFlairs[flairIndex];
+      if (!Flairs.canUserApply(userRefs, applicationFlair, Flairs.getUserFlairs(req.user, allFlairs))) {
```
We could probably add a wrapper function inside Flairs service that we can pass the name into rather than having to do the grabbing of the actual flair out here, but it's not important.

YaManicKill commented on api/controllers/FlairController.js:
```diff
@@ -5,122 +5,65 @@ var _ = require("lodash");
+  denyApp: async function (req, res) {
+    try {
+      var matching_apps = await Application.destroy(req.allParams().id);
+      var apps = await Flairs.getApps();
```
Why are we returning all of the apps rather than the one that was deleted?

not-an-aardvark replied:
I made it so the app list auto-refreshes whenever one approves or denies an app -- that way it's less likely that two people will work on an app at the same time.

YaManicKill commented on api/controllers/FlairController.js:
```diff
@@ -5,122 +5,65 @@ var _ = require("lodash");
+      var pmContent = 'Your application for ' + Flairs.formattedName(app.flair) + ' flair on /r/' + app.sub + ' has been approved.';
+      promises.push(Reddit.sendPrivateMessage(refreshToken, 'FlairHQ Notification', pmContent, user.name));
+      promises.push(Application.destroy({id: req.allParams().id}));
+      await* promises;
```
Is this the equivalent of .all? Nice.

not-an-aardvark replied:
Yep. `await* promises` === `await Promise.all(promises)`

YaManicKill commented on api/controllers/FlairController.js:
```diff
@@ -5,122 +5,65 @@ var _ = require("lodash");
-      });
-    });
+      var user = await User.findOne(app.user),
+        shortened = app.sub === 'pokemontrades' ? 'ptrades' : 'svex';
```
Can you remove the comma from this and just define it seperately. Mainly because the await makes it confusing to look at with that.

not-an-aardvark replied:
Sure, will fix.

YaManicKill commented on api/controllers/FlairController.js:
```diff
@@ -5,122 +5,65 @@ var _ = require("lodash");
+      await* promises;
+      return res.ok(await Flairs.getApps());
+    } catch (err) {
+      return res.serverError(err);
```
Maybe can we have multiple errors. 1 for if the flair isn't set correctly (and send that so we get the error and we can show it) and 1 for everything else (that we can sanitise if we want).

not-an-aardvark replied:
We already do the validity check [here](https://github.com/pokemontrades/flairhq/blob/refactor-client-logic/api/controllers/FlairController.js#L71) and throw a verbose error if the user messed up. This try-catch is after we've validated the input, so stuff can only go wrong on our end.

YaManicKill commented on api/services/Flairs.js:
```diff
@@ -221,6 +221,31 @@ exports.flairCheck = function (ptrades, svex) {
   return response;
 };
 
+exports.makeNewCSSClass = function (previous_flair, new_addition, subreddit) {
```
Do you have tests for this?

not-an-aardvark replied:
Yep, sorry about the hard-to-read regex golf. There are a bunch of tests for it though:

```
  Upgrading/combining flairs
    pokemontrades flairs
      ✓ pokeball + premierball → premierball
      ✓ greatball hok + ultraball → ultraball hok
      ✓ ovalcharm1 + shinycharm → shinycharm1
      ✓ greatball + involvement → greatball1
      ✓ ovalcharm + involvement → ovalcharm1
      ✓ premierball hok + involvement → premierball1 hok
      ✓ (no flair) + banned → banned
      ✓ pokeball + banned → pokeball banned
      ✓ pokeball1 + banned → pokeball1 banned
      ✓ ovalcharm hok + banned → ovalcharm banned
      ✓ ovalcharm1 hok + banned → ovalcharm1 banned
      ✓ masterball1 hok something-else + banned → masterball1 banned
      ✓ dreamball banned + banned → dreamball banned
      ✓ banned + banned → banned
    SVExchange flairs
      ✓ (no flair) + lucky → lucky
      ✓ lucky + egg → egg
      ✓ eevee + togepi → togepi
      ✓ togepi smartribbon + torchic → torchic smartribbon
      ✓ (no flair) + cuteribbon → cuteribbon
      ✓ cuteribbon + coolribbon → coolribbon
      ✓ manaphy smartribbon + toughribbon → manaphy toughribbon
      ✓ eggcup + beautyribbon → eggcup beautyribbon
      ✓ beautyribbon + eggcup → eggcup beautyribbon
      ✓ (no flair) + banned → banned
      ✓ lucky + banned → lucky banned
      ✓ lucky cuteribbon + banned → lucky cuteribbon banned
      ✓ toughribbon + banned → toughribbon banned
      ✓ banned + banned → banned
      ✓ lucky banned + banned → lucky banned
      ✓ lucky cuteribbon banned + banned → lucky cuteribbon banned
      ✓ toughribbon banned + banned → toughribbon banned
      ✓ pichu banned + toughribbon → pichu toughribbon banned
      ✓ coolribbon banned + togepi → togepi coolribbon banned
```

YaManicKill commented on api/services/References.js:
```diff
@@ -85,59 +88,57 @@ exports.getRedditUser = function (username) {
   }
 };
-exports.numberOfPokemonGivenAway = function (user) {
+exports.numberOfPokemonGivenAway = function (refs) {
```
Ha, good call.

YaManicKill commented on assets/adminCtrl.js:
```diff
@@ -10,43 +7,26 @@ module.exports = function ($scope, io) {
-        $scope.$apply();
+        $scope.flairApps = data;
+      } else if (res.statusCode === 404) {
+        $scope.flairApps = data;
```
Oh I see. Whatever, I guess this can cause less issues.

YaManicKill commented on assets/app.js:
```diff
@@ -28,19 +28,6 @@ var fapp = ng.module("fapp", [
   'fapp.md'
 ]);
 
-fapp.factory('UserFactory', function () {
```
Why?

not-an-aardvark replied:
We don't need this anymore since `$scope.user` is created synchronously.

YaManicKill commented on assets/indexCtrl.js:
```diff
@@ -114,7 +107,7 @@ module.exports = function ($scope, io) {
           $scope.addRefError = data.err;
         } else {
-          $scope.addRefError = "Already added that URL.";
+          $scope.addRefError = "There was an error.";
```
Ah yes, thanks. We should probably give some more useful errors than this, but at least this isn't misleading.

not-an-aardvark replied:
The server actually sends 'Already added that URL' in `data.err` if the URL was already added, so this just prevents the misleading messages.

YaManicKill commented on assets/refCtrl.js:
```diff
@@ -87,9 +66,6 @@ module.exports = function ($scope, $filter, io) {
   $scope.modSaveProfile = function () {
     $scope.ok.modSaveProfile = false;
     $scope.spin.modSaveProfile = true;
-    if (!$scope.user.isMod) {
```
Why is this removed? Is there a check elsewhere?

not-an-aardvark replied:
The buttons don't appear if the user is not a mod. This check is clientside anyway, so it didn't have any security value.

YaManicKill commented on assets/search/header.ejs:
```diff
@@ -1,6 +1,5 @@
-<form class="navbar-form navbar-left dropdown hidden"
```
This will be shown on the search page now, won't it?

not-an-aardvark replied:
No, there's a separate check for that.

YaManicKill commented on assets/search/search.controller.js:
```diff
@@ -3,10 +3,10 @@ var io = require("sails.io.js")(socket);
   var vm = this;
 
-  vm.user = {};
+  vm.user = $scope.user;
```
Oh, how does this work?

not-an-aardvark replied:
`user` is one of the EJS variables that gets injected in `layout.ejs`.

YaManicKill commented on assets/styles/importer.less:
```diff
@@ -184,6 +184,14 @@ td h3 {
   color: green;
 }
 
+[ng\:cloak], [ng-cloak], .ng-cloak {
```
Did you not figure out a way to make the pokeball keep going till everything was sorted in angular?

not-an-aardvark replied:
[It's apparently possible but a bit messy](https://stackoverflow.com/questions/14968690/sending-event-when-angular-js-finished-loading). As an alternative, we could exclude the pokeball from the ng-cloak; that way the pokeball will start spinning before the angular stuff is sorted.

YaManicKill replied:
Hmmm, I'll merge this in and spend a bit of time trying ot make it look a bit nicer.

YaManicKill commented on assets/userCtrl.js:
```diff
@@ -96,115 +97,12 @@ module.exports = function ($scope, $filter, $location, UserFactory, io) {
-    } else {
-      window.setTimeout($scope.getReferences, 1000);
-    }
-  };
```
Beautiful.

YaManicKill commented on assets/userCtrl.js:
```diff
@@ -96,115 +97,12 @@ module.exports = function ($scope, $filter, $location, UserFactory, io) {
-              }
-
-              $scope.user.flairGames = [];
-              for (var j = 0; j < games.length || j < igns.length || j < tsvs.length; j++) {
```
Where has this stuff gone?

not-an-aardvark replied:
Vanished, apparently. I'll put it back.

YaManicKill commented on assets/userCtrl.js:
```diff
@@ -477,5 +356,170 @@ module.exports = function ($scope, $filter, $location, UserFactory, io) {
+  };
+  $(document).ready(function () {
+    $scope.loaded = true;
+  });
```
And just when this file was a decent size.

YaManicKill commented on assets/views/home/header.ejs:
```diff
@@ -33,6 +33,7 @@
             <li><a href="#" data-toggle="modal" data-target="#flairAppList">Flair App List</a></li>
             <li><a href="#" data-toggle="modal" data-target="#flairMod">Edit Flairs</a></li>
             <li><a href="/mod/banuser">Ban a User</a></li>
+            <li><a target="_self" href="/mod/downloadModmail" download="modmails.txt">Download Modmails</a></li>
```
Why have we got this here?

I had thought about adding it as an option when the mod is searching. so show all the results, and also do a quick search through this .txt file and allow them to download it if it has some results in it.

Means less steps for the mod.

not-an-aardvark replied:
I thought it would be nice to have a link in case people want to download the entire archive. That way we can take the dropbox link offline, so we don't have to worry about people accidentally leaking it.

Keep in mind that this button doesn't the old, hard-to-parse txt file - it downloads a a file in JSON format directly from our database.

Also, I'm not sure why the other thing would be less steps -- this is a one-click download from the header.

YaManicKill replied:
What do you want the modmail for? To search something, yes?

If you download the entire file, what would be the purpose of that? It is less steps, because the user isn't downloading something if they are just searching. It's ... unnecessary. I just don't see why we would need the file at all now.

not-an-aardvark replied:
¯\\\_(ツ)_/¯

I guess it might be useful if people want to do some very specific type of search/data collection/statistical analysis that we don't provide. I think we should offer the entire file for download somewhere, although I'd be fine with putting it in a less prominent location.

YaManicKill replied:
Meh, fine. I think it's silly, but whatever.

YaManicKill commented on assets/views/home/index.ejs:
```diff
@@ -104,7 +104,7 @@
       </form>
 
-      <%- partial('references.ejs', {theUser: "user", refPage: false}) %>
+      <%- partial('references.ejs', {theUser: "refUser", refPage: false}) %>
```
Well, "theUser" is unnecessary if "refUser" is always going to be the user we use now.

YaManicKill commented on assets/views/home/references.ejs:
```diff
@@ -405,23 +196,7 @@
-          <span class="glyphicon glyphicon-trash"></span>
-        </a>
-      </td>
+      <%- partial('row.ejs') %>
```
\o/ Now all we need is to collapse it more and have an ng-repeat for the sections as well. But that's a great start.

YaManicKill commented on assets/views/home/row.ejs:
```diff
@@ -0,0 +1,44 @@
+<td><strong>{{$index + 1}}.</strong></td>
```
Can we space this out a bit, it's quite hard to read currently.

YaManicKill commented:
Couple of small things, mainly the login stuff and sharedClientFunctions.

Fix them and I'll merge. 

And no more humungous PRs. Small ones from now on :-P

not-an-aardvark commented:
Fixed a few things. Also, I fixed [this bug](https://i.gyazo.com/db9bde303f0b05076d8f2a428654fc04.png) where details on a second game wasn't appearing by default. (Somehow I never consciously acknowledged it for long enough to report it.)

YaManicKill commented:
Ok, I think I'm happy apart from the shared client functions. Fix that, and we're good.

not-an-aardvark commented:
I put the shared function thing back as it was in version 2.2
